chapter05/src/main/java/com/seaofnodes/simple/node/ScopeNode.java

```diff
@@ -86,23 +95,30 @@ public Node update(String name, Node n) {
     }
 
     /**
-     * Duplicate a ScopeNode; including all levels.
+     * Duplicate a ScopeNode; including all levels, up to Nodes.  So this is
+     * neither shallow (would dup the Scope but not the internal HashMap
+     * tables), nor deep (would dup the Scope, the HashMap tables, but then the
+     * also program Nodes).
      *
-     * We do it in a way that ensures that the new ScopeNode becomes
-     * additional user of all defined names
+     * The new Scope is a full-fledged Node with proper use<->def edges.
      */
     public ScopeNode dup() {
-        ScopeNode dupScope = new ScopeNode();
-        for (HashMap<String, Integer> tab: _scopes) {
-            dupScope.push(); // Create same level in the duplicate
-            for (Map.Entry<String, Integer> e: tab.entrySet()) {
-                String name = e.getKey();
-                Integer idx = e.getValue();
-                Node n = in(idx);
-                dupScope.define(name, n); // Ensure that dupScope is a user of n
-            }
+        ScopeNode dup = new ScopeNode();
+        for( HashMap<String, Integer> tab : _scopes ) {
+            dup.push();    // Create same level in the duplicate
+            for( String name : tab.keySet() )
```

Comment:
The order of `keySet` is not the order of insertion and so might mix up the indices for merging.

The following program returns `Phi(Region15,4,2)`, but `2` is never assigned to `c`.
```c
int c = 3;
int b = 2;
if (arg == 1) {
    b = 3;
    c = 4;
}
return c;
```

Reply:
The version I wrote didn't assume indices were in same order in the dup, but with the new approach, we have to make sure I guess